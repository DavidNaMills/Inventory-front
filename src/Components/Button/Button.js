import React from 'react';
import classes from './button.module.scss';

const allTypes = {
    base: {
        base: classes.button__base,
        shadow: classes.button__base_shadow,
        border: classes.button__base_border,
        skeleton: classes.button__base_skeleton,
    },

    submit: {
        base: classes.button__submit,
        shadow: classes.button__submit_shadow,
        border: classes.button__submit_border,
        skeleton: classes.button__submit_skeleton,
    },

    clear: {
        base: classes.button__clear,
        shadow: classes.button__clear_shadow,
        border: classes.button__clear_border,
        skeleton: classes.button__clear_skeleton,
    },

    danger: {
        base: classes.button__danger,
        shadow: classes.button__danger_shadow,
        border: classes.button__danger_border,
        skeleton: classes.button__danger_skeleton,
    },

    information: {
        base: classes.button__information,
        shadow: classes.button__information_shadow,
        border: classes.button__information_border,
        skeleton: classes.button__information_skeleton,
    },
};


const Button = ({ config = {}, onClick, label = null, skeleton = false, raised = false, full = false, type = 'base', border = false, bold = false, wide = false, isDisabled = false, Icon = null, noStyle=null }) => {
    const fullStyle = [classes.button];
    let which = type;
    const style = {};

    if (isDisabled) {
        fullStyle.push(classes.button__disabled);
    } else if(noStyle) { fullStyle.push(classes.button_noStyle);
    } else {
        if (allTypes[type]) {
            which = type;
            fullStyle.push(allTypes[type].base);
        } else {
            which = 'base';
            fullStyle.push(allTypes['base'].base);
        }

        if (skeleton) { fullStyle.push(allTypes[which].skeleton); }
        if (raised) { fullStyle.push(allTypes[which].shadow); }
        if (border) { fullStyle.push(allTypes[which].border); }
    }
    if (full) { fullStyle.push(classes.button_full); }
    if (bold) { fullStyle.push(classes.button_bold); }
    
    
    if (wide) {
        style.width = '30%';
    }

    return (
        <button
            {...config}
            onClick={onClick}
            style={style}
            className={fullStyle.join(' ')}
            disabled={isDisabled}
        ><p>{label}</p>{Icon && <Icon />}</button>
    )
}

export default React.memo(Button);