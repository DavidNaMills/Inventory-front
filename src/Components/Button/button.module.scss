@import '../../styles/mixins/buttonMixins.module.scss';
@import '../../styles/variables/colors.module.scss';

.button{
    display: flex;
    align-content: center;
    justify-content: center;
    font-size: 16px;
    border: none;
    padding: 10px;
    padding-left: 20px;
    padding-right: 20px;
    border-radius: 2px;

    p{
        padding-top: 5px;
        padding-right: 10px;
    }

    &__disabled{
        background-color: $disabled-back;
        color: $disabled-color;
    }

    &__base{
        @include buttonStyle($base-back, $base-color);
        &:hover{
            @include onHover($base-shadow, $light-text, $base-back);
        }

        &_shadow{@include buttonShadow($base-shadow);}
        &_border{@include buildBorder($base-color);}
        &_skeleton{@include buildSkeleton($base-back, $base-color); }
    };
    
    &__submit{
        @include buttonStyle($submit-back, $submit-color);
        &:hover{
            @include onHover($submit-shadow, $light-text, $submit-back);
        }

        &_shadow{ @include buttonShadow($submit-shadow); }
        &_border{ @include buildBorder($submit-color); }
        &_skeleton{@include buildSkeleton($submit-back, $submit-color); }
    }
    
    &__clear{
        @include buttonStyle($clear-back, $clear-color);
        &:hover{
            @include onHover($clear-shadow, $clear-color, $clear-back);
        }

        &_shadow{ @include buttonShadow($clear-shadow); }
        &_border{ @include buildBorder($clear-color); }
        &_skeleton{@include buildSkeleton($clear-back, $clear-color); }
    }
    
    &__danger{
        @include buttonStyle($danger-back, $danger-color);
        &:hover{
            @include onHover($danger-shadow, $light-text, $danger-back);
        }
        &_shadow{ @include buttonShadow($danger-shadow); }
        &_border{ @include buildBorder($danger-color); }
        &_skeleton{@include buildSkeleton($danger-back, $danger-color); }
    }
    
    &__information{
        @include buttonStyle($information-back, $information-color);
        &:hover{
            @include onHover($information-shadow, $light-text, $information-back);
        }
        &_shadow{ @include buttonShadow($information-shadow); }
        &_border{ @include buildBorder($information-color); }
        &_skeleton{@include buildSkeleton($information-back, $information-color); }
    }

    &_flat{
        border: none;
        box-shadow: none;
        -webkit-box-shadow: none;
        -moz-box-shadow: none;
    }

    &_full{
        width: 100%;
    }

    &_bold{
        font-weight: bold;
    }

    &_noStyle{
        background-color: transparent;
        border: none;
        box-shadow: none;

        &:hover{
            background-color: rgb(212, 212, 212);
            padding-top: 10px;
            padding-bottom: 10px;
            padding-left: 5px;
            padding-right: 10px;
            border-radius: 100px;
        }
    }
}