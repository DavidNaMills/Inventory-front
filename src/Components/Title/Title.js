import React from 'react';
import classes from './title.module.scss';


const Title = (props) => {
    const { size = null, message = '', center = false } = props;
    let element = null;

    switch (size) {
        case 'h1':
            element = <h1 className={[classes.title_h1, center ? classes.title__center : null].join(' ')} >{message}</h1>
            break;
        case 'h2':
            element = <h2 className={[classes.title_h2, center ? classes.title__center : null].join(' ')}>{message}</h2>
            break;
        case 'h3':
            element = <h3 className={[classes.title_h3, center ? classes.title__center : null].join(' ')}>{message}</h3>
            break;
        case 'h4':
            element = <h4 className={[classes.title_h4, center ? classes.title__center : null].join(' ')}>{message}</h4>
            break;
        case 'h5':
            element = <h5 className={[classes.title_h5, center ? classes.title__center : null].join(' ')}>{message}</h5>
            break;
        case 'h6':
            element = <h6 className={[classes.title_h6, center ? classes.title__center : null].join(' ')}>{message}</h6>
            break;
        default:
            element = <h2 className={[classes.title_h2, center ? classes.title__center : null].join(' ')}>{message}</h2>
            break;
    };

    return (
        <div>
            {element}
        </div>
    )
}

export default React.memo(Title);