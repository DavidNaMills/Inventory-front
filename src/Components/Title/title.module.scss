@import '../../styles/variables/colors.module.scss';
@import '../../styles/fonts/fontSizes.module.scss';
@import '../../styles/mixins/title.module.scss';

.title {
    @include setTitle($h2, $title-color);
    
    &_h1{
        @include setTitle($h1, $title-color);
    };

    &_h2{
        @include setTitle($h2, $title-color);
    };
    
    &_h3{
        @include setTitle($h3, $title-color);
    }

    &_h4{
        @include setTitle($h4, $title-color);
    }

    &_h5{
        @include setTitle($h5, $title-color);
    }

    &_h6{
        @include setTitle($h6, $title-color);
    }

    &__center{
        text-align: center;
    }


}