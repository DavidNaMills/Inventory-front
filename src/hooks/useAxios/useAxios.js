import {useState} from 'react';

const defaultState = {
    isLoading: false,
    errors: null,
    data: null,
    id: null
}

const useAxios = (axios) =>{
    const [fetchState, setFetchState] = useState(defaultState);

    const makeCall = ({url, method, data, callBack=()=>{}, newId=null}) =>{
        const fullConfig = {
            url,
            method
        };
        if(data){
            fullConfig.data = data
        }
        setFetchState({...defaultState, isLoading: true, id: newId})

        return axios(fullConfig)
        .then(res=>{
            // setTimeout(()=>{
                const result = {
                    id: newId,
                    isLoading: false,
                    errors: null,
                    data: res.data
                }

                setFetchState(result)
                callBack(res.data);
                return result;
            // }, 3000);
        })
        .catch(err=>{
            setFetchState({...defaultState, id: newId});
            callBack({
                id: newId,
                error: true,
                message: 'went wrong',
                data: null
            });
        })
    }

    return {
        fetchState,
        makeCall
    }
}

export default useAxios;