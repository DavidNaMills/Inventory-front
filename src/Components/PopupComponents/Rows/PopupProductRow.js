import React from 'react';
import { isMobileOnly } from 'react-device-detect';
import ProductRow from '../../TableDisplay/Row/ProductRow/ProductRow';
import classes from '../Popup.module.scss';
import Title from '../../Title/Title';
import {useLocale} from '../../../Context/LocaleContext';


const config2 = { main: ['', 'name', 'qtyInStock'] };

const PopupProductRow = (props) => {
    const { prods, isPortrait=true } = props;
    const locale = useLocale();

    const tempScrollableProd = {
        overflowY: 'auto',
        maxHeight: isPortrait ? '30vh' : '15vw',
    }

    return (
        prods.length > 0
            ? <div className={classes.popup__color2}>
                <div className={classes.popup__mediumSpacing}>
                    <Title message={`${locale['products_msg']}: ${prods.length>0 ? prods.length : 0} ${locale['in_ttl']}`} size='h4' />
                </div>
                <div style={tempScrollableProd}>
                    {
                        prods.map((c, d) =>
                            <div key={d} className={d % 2 ? classes.popup__row_odd : classes.popup__row_even}>
                                <ProductRow
                                    data={c}
                                    config={config2}
                                    withTitle={false}
                                    withImage={isMobileOnly ? false : true}
                                    stack={isMobileOnly ? true : false}
                                />
                            </div>
                        )
                    }
                </div>
            </div>
            : <p className={classes.popup__mediumSpacing}>{locale['no_products_type']}</p>
    )
}

export default PopupProductRow;