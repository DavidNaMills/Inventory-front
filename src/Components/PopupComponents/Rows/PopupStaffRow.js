import React from 'react';
import { isMobileOnly, isMobile } from 'react-device-detect';
import StaffTypeRow from '../../TableDisplay/Row/StaffTypeRow';
import classes from '../Popup.module.scss';
import Title from '../../Title/Title';
import {useLocale} from '../../../Context/LocaleContext';

const config = {main: ['', 'name', 'role']};

const PopupStaffRow = (props) => {
    const { staff, isPortrait } = props;
    const locale = useLocale();

    const tempScrollableStaff = {
        overflowY: 'auto',
        height: isMobile ? isPortrait ? '20vh' : '10vw' : '15vh',
    }
    
    return (
        staff.length > 0 ? (
            <div className={classes.popup__color1}>
                <div className={classes.popup__mediumSpacing}>
                    <Title message={staff.length>0 ? `${locale['staff_count']}: ${staff.length} ${locale['in_ttl']}` : `${locale['no_staff_type']}`} size='h4' />
                </div>
                <div style={tempScrollableStaff}>
                    {
                        staff.map((c, d) =>
                            <div key={d} className={d % 2 ? classes.popup__row_odd : classes.popup__row_even}>
                                <StaffTypeRow
                                    config={config}
                                    data={c}
                                    withTitle={false}
                                    isMobileOnly={isMobileOnly}
                                />
                            </div>
                        )
                    }
                </div>
            </div>
        )
            : <p className={classes.popup__mediumSpacing}>No staff at this location</p>
    )
}

export default PopupStaffRow;