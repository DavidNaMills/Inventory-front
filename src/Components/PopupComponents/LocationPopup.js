import React, { useEffect, useState } from 'react';
import classes from './Popup.module.scss';
import Title from '../Title/Title';
import { withOrientationChange } from 'react-device-detect';
import useAxios from '../../hooks/useAxios/useAxios';
import axios from '../../Axios/axiosInstance';
import { URLS } from '../../Axios/urls';
import Button from '../Button/Button';

import PopupProductRow from './Rows/PopupProductRow';
import PopupStaffRow from './Rows/PopupStaffRow';
import Spinner from '../../Components/Spinner/Spinner';
import {useLocale} from '../../Context/LocaleContext';


const LocationPopup = (props) => {
    const { item, onClose, isPortrait } = props;
    const [staff, setStaff] = useState([]);
    const [prods, setProds] = useState([]);
    const { fetchState, makeCall } = useAxios(axios);
    const locale = useLocale();

    const initPopup = async () =>{
        const t = makeCall({
            url: `${URLS.allProducts}/${item._id}/0/0`,
            method: 'GET',
            newId: 'products'
        });
    
        const d = makeCall({
            url: `${URLS.staffLocations}/${item._id}`,
            method: 'GET',
            newId: 'staff'
        });

        Promise.all([t, d])
        .then(res=>{
            if(!res[0].errors && !res[1].errors){
                setProds(res[0].data.products);
                setStaff(res[1].data.data);
            }
        })
    }

    useEffect(() => {
        initPopup();
    }, [])

    return (
        <div className={classes.popup}>
            <Title message={item.displayValue} size='h2' center />
            {
                item.description &&
                <p className={classes.popup__mediumSpacing}>{item.description}</p>
            }

            {fetchState.isLoading
                ? <Spinner withBackground={false}/>
                : <div>
                    <PopupStaffRow staff={staff} isPortrait={isPortrait}/>
                    <PopupProductRow prods={prods} isPortrait={isPortrait}/>
                </div>
            }
            <Button onClick={onClose} label={locale['close']} full />
        </div>
    )
}

export default withOrientationChange(LocationPopup);