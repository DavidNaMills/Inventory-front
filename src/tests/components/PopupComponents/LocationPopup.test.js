import React from 'react';
import '../../testHelpers/context';
import renderer from 'react-test-renderer';
import {mount} from 'enzyme';

import LocationPopup from '../../../Components/PopupComponents/LocationPopup';

import PopupProductRow from '../../../Components/PopupComponents/Rows/PopupProductRow';
import PopupStaffRow from '../../../Components/PopupComponents/Rows/PopupStaffRow';
import Button from '../../../Components/Button/Button';
import Title from '../../../Components/Title/Title';

jest.mock('../../../hooks/useAxios/useAxios');
import useAxios from '../../../hooks/useAxios/useAxios';

// jest.mock('react-device-detect');
import {withOrientationChange} from 'react-device-detect';

const mockClose = jest.fn();
const mockCall = jest.fn();

const item = {
    _id: '123',
    displayValue: 'Home',
    description: 'this is a description'
};

const mockFetchState = {
    isLoading: false,
    errors: null,
    data: {
        data: [
            {name: 'test1'}
        ],
        products: [
            { _id: '2121', name: 'cups', qtyInstock: '123' },
            { _id: '7894512', name: 'plates', qtyInstock: '789' },
        ]
    },
    id: null
}


describe('<LocationPopup/> component test suite', ()=>{
    describe('all options true', ()=>{
        let wrapper;

        beforeEach(()=>{
            useAxios.mockImplementation(()=>({
                fetchState: mockFetchState,
                makeCall: mockCall
            }));

            jest.mock('react-device-detect', ()=>({
                withOrientationChange: (component)=>{
                    component.defaultProps = {
                        ...component.defaultProps,
                        isPortrait: true
                    };
                    return component;
                }
            }))

            wrapper = mount(<LocationPopup item={item} onClose={mockClose}/>);
        })

        it('matches snapshot', ()=>{
            const comp = renderer.create(<LocationPopup item={item} onClose={mockClose}/>);
            const app = comp.toJSON();
            expect(app).toMatchSnapshot();
        });
        
        it('exists', ()=>{
            expect(wrapper.exists).toBeTruthy();
        });
        
        it('renders a <div> container with classname "popup"', ()=>{
            const div = wrapper.find('div');
            expect(div.length).toBe(3);
            expect(div.at(0).props()).toHaveProperty('className', 'popup');
        });
        
        it('renders a <Title/> component', ()=>{
            const ttl = wrapper.find(Title);
            expect(ttl.length).toBe(1);
            expect(ttl.props()).toHaveProperty('message', item.displayValue);
            expect(ttl.props()).toHaveProperty('size', 'h2');
            expect(ttl.props()).toHaveProperty('center', true);
        });
        
        it('renders a <p> element with className "popup__mediumSpacing" with the product description', ()=>{
            const p = wrapper.find('p');
            expect(p.length).toBe(4);
            expect(p.at(0).props()).toHaveProperty('className', 'popup__mediumSpacing');
            expect(p.at(0).text()).toEqual(item.description);
        });
        
        it('renders a <PopupStaffRow/> and <PopupProductRow/> components', ()=>{
            const stff = wrapper.find(PopupStaffRow);
            expect(stff.length).toBe(1);
            expect(stff.props()).toHaveProperty('staff', mockFetchState.data.data)
        });
        
        it('renders a <Button/> component', ()=>{});
    });

    describe('alternative conditions tested', ()=>{
        it('does not render a <p> element if description is null', ()=>{});
    });
});