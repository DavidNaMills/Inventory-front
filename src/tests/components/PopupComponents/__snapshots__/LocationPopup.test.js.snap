// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LocationPopup/> component test suite all options true matches snapshot 1`] = `
<div
  className="popup"
  style={
    Object {
      "maxHeight": "90%",
      "maxWidth": "100%",
    }
  }
>
  <div>
    <h2
      className="title_h2 title__center"
    >
      Home
    </h2>
  </div>
  <p
    className="popup__mediumSpacing"
  >
    this is a description
  </p>
  <div>
    <p
      className="popup__mediumSpacing"
    >
      No staff at this location
    </p>
    <p
      className="popup__mediumSpacing"
    >
      No products of this type
    </p>
  </div>
  <button
    className="button button__base button_full"
    disabled={false}
    onClick={[MockFunction]}
    style={Object {}}
  >
    <p>
      Close
    </p>
  </button>
</div>
`;
